### ISSUE TITLE
Numerical issues with `BFloat16`

### ISSUE THREAD

**Issue by AntonOresten** [FluxML/Optimisers.jl#215]
I've been trying to do training in BFloat16, and I just figured out the primary cause for NaNs in the backward pass, namely that momentum parameters in the optimisers state need to be stored with higher precision than BFloat16 offers.

For example, the default beta values for Adam are (0.9, 0.999). The second value can't be stored in BFloat16, as the nearest values are 0.996 and 1.0, so it gets rounded to 1.0.

I think the simplest solution is to store momentum in Float32, but convert to BFloat16 as part of the broadcasting:

```julia
function apply!(o::Adam, state, x::AbstractArray{T}, dx) where T
    η, β, ϵ = T(o.eta), o.beta, T(o.epsilon)
    mt, vt, βt = state

    @.. mt = T(β[1] * mt + (1 - β[1]) * dx)
    @.. vt = T(β[2] * vt + (1 - β[2]) * abs2(dx))
    dx′ = @lazy mt / (1 - βt[1]) / (sqrt(vt / (1 - βt[2])) + ϵ) * η

    return (mt, vt, βt .* β), dx′
end
```

Conversion to BFloat16 would also need to be avoided in the state init.

I will create a PR for this. It's not clear if this should also be done with Float16 (might be breaking). There might also be many optimisers that suffer from BFloat16 numerical issues.

EDIT: it's likely that 0.999 is simply not used for BFloat16.

**Comment by mcabbott** [comment]
That sounds fine to me. The current code using `eltype(x)` is trying hard to avoid ever doing broadcasts in Float64 accidentally, with Float32 parameters, as this was once an easy silent performance bug.

Does this need to be configurable for Float16? It would certainly be tidiest not to be... if there was just a fixed mapping from parameter eltype S -> momentum and computation eltype T, something like: 
```
function apply!(o::Adam, state, x::AbstractArray{S}, dx) where S
    T = enlarge(S)
    η, β, ϵ = T(o.eta), o.beta, T(o.epsilon)
```
and something like making `elargedzero(x)` called by `init`.

**Comment by AntonOresten** [comment]
I think the main issue in this case stems from inadequate precision when dealing with numbers just below 1, and then dividing by 1 minus that. This doesn't mean we need to store any states in higher precision, but maybe that is the robust and more general solution.

There can be significant gains to having heavier optimizers (e.g. #203) in BFloat16 still, when the apply function is more than broadcasting a couple of arrays, and benefits from high-performing GEMMs.

**Comment by CarloLucibello** [comment]
@AntonOresten It would be ok for you to cast the gradient to Float32 before the update?

I suggest the following change to all rules:
- all hyperparams in the state are stored as Float64
- in `apply!`, hyperparams are cast to the widest type between x and dx, that is `T = promote_type(eltype(x), eltype(dx)`
- computation is performed in precision `T`.
- The type of momentum etc... states remain the same as it is now, that is given by the types of the array in the model at setup type: `opt_state = Optimisers.setup(Adam(), model)`.

This way seems to me that we have full flexibility on how we want to perform the computation by casting the model in setup and/or by casting the gradient.

**Comment by CarloLucibello** [comment]
Notice that the most common usage pattern for working with 16-bit precision is mixed precision training, where the following happens:
- There is a master copy of the weight in fp32
- The forward and backward run in fp16 or bf16:
   - This is done either by keeping in memory a low-precision copy or by having each operation downcast to 16 bits. The latter is what is done with the `torch.autocast` context, so that there is no need to maintain an extra copy of the model.
  -  Some operations, in particular those involving reductions (e.g. losses and normalization layers) and softmax (attention) are blacklisted and run in fp32
   - For fp16 also scaling up the loss is needed, not for bf16 though.
- Gradients are then cast to fp32 (and then rescaled in the case of fp16)
- Optimizer states are kept in fp32 usually
- The fp32 weights are updated

I wrote this down also as a reference for myself, since I'm playing with integrating mixed precision training in flux.
